Installing opencv-python

Hey,

Is there a way to install opencv-python?
The standard
`pip3 install opencv-python`
returns 
```
Could not find version that satisfies the requirement opencv-python (from versions: )
No matching distribution found for opencv-python
```

@its-pointless I tried building opencv again, with `LDFLAGS=" -llog -lboost_python-3" cmake -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_INSTALL_PREFIX=$PREFIX -DBUILD_opencv_python3=on -DBUILD_opencv_python2=off -DWITH_QT=OFF -DWITH_GTK=OFF ..`

But ran into quite a lot of undefined references 
```
/data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_FromFormat'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_SetItem'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyList_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_AsWideChar'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceFloorDivide'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Size'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyModule_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyArg_ParseTupleAndKeywords'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Items'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyProperty_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyMethod_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyList_Insert'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_ReferenceError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyFloat_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyTuple_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Lshift'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_ClearWeakRefs'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_CallMethod'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyMem_Free'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyCFunction_NewEx'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceAdd'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_AsUnsignedLongLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Add'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_GetAttrString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_AsUTF8String'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_SetAttrString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_WarnEx'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_NoMemory'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_SetObject'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyIter_Next'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Multiply'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `_Py_NotImplementedStruct'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_RichCompare'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyRun_FileExFlags'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_RuntimeError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_SetString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_IsInstance'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_DelItem'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyMethod_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_GetItem'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_ValueError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_TypeError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyStaticMethod_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyTuple_GetItem'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyComplex_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_ExceptionMatches'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceAnd'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyType_Ready'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_FromLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyComplex_ImagAsDouble'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_AsSsize_t'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyBytes_Size'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_GetItem'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_NewException'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_Clear'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyBool_FromLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyList_Append'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Clear'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Update'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Or'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_And'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_CallFunction'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyTuple_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_OverflowError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Remainder'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyType_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_SetAttr'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_Occurred'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyModule_Create2'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_AsLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyImport_ImportModule'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyEval_GetGlobals'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceOr'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceXor'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `_Py_fopen'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyRun_StringFlags'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_GetItemString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyList_Reverse'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyList_Sort'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_Size'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `_Py_NoneStruct'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_IsTrue'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `_PyType_Lookup'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyEval_CallFunction'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyType_GenericAlloc'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceRshift'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyStaticMethod_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyWeakref_NewRef'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Rshift'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceRemainder'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_IndexError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyBool_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyCallable_Check'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_AsUnsignedLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyBaseObject_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_FromUnsignedLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_AsUTF8'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Keys'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `_PyObject_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_FromString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_InternFromString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_AttributeError'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Copy'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyBytes_AsString'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyExc_StopIteration'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_Call'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyLong_AsLongLong'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyType_IsSubtype'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyTuple_Size'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceLshift'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyErr_Format'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PySlice_New'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_FromStringAndSize'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceMultiply'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyObject_GetAttr'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyCFunction_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyUnicode_FromEncodedObject'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyComplex_RealAsDouble'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyMem_Malloc'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Subtract'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_Xor'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_InPlaceSubtract'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyList_Type'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyNumber_FloorDivide'
    /data/data/com.termux/files/usr/bin/../lib/libboost_python-3.so: undefined reference to `PyDict_Values'
```
Could/should this be dealt with when compiling boost or should/do I need to link against the corresponding libs on device?

Hi! I've been trying to reproduce the OpenCV compiling in my Termux (Android 8, aarch64) and I'm repeteadly failing at this point: 

> [ 29%] Building CXX object modules/core/CMakeFiles/opencv_core.dir/src/system.cpp.o       /data/data/com.termux/files/home/opencv-3.0.0/modules/core/src/system.cpp:645:29: error:        use of undeclared identifier                 'ANDROID_LOG_ERROR'                            __android_log_print(ANDROID_L...                                 ^                1 error generated.
make[2]: * [modules/core/CMakeFiles/opencv_core.dir/build.make:1026: modules/core/CMakeFiles/opencv_core.dir/src/system.cpp.o] Error 1                                            make[1]: * [CMakeFiles/Makefile2:1585: modules/core/CMakeFiles/opencv_core.dir/all] Error 2

Using as Cmake flags `LDFLAGS=" -llog -lpython3"`. I've readed a lot and I supose it's related with the uncorrect linking something, but I'm not a CMake expert. I could use some help. 

Thank you, 

Isma

Now this.

CMake Error at /data/data/com.termux/files/usr/share/cmake-3.12/Modules/CMakeTestCXXCompiler.cmake:45 (message):
  The C++ compiler

    "/data/data/com.termux/files/usr/bin/c++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /data/data/com.termux/files/home/opencv/build/CMakeFiles/CMakeTmp

    Run Build Command:"/data/data/com.termux/files/usr/bin/ninja" "cmTC_246f2"
    [1/2] Building CXX object CMakeFiles/cmTC_246f2.dir/testCXXCompiler.cxx.o
    /data/data/com.termux/files/usr/bin/arm-linux-androideabi-ld: cannot find -lboost_python
    clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
    [2/2] Linking CXX executable cmTC_246f2
    FAILED: cmTC_246f2
    : && /data/data/com.termux/files/usr/bin/c++    -llog -lboost_python   -rdynamic CMakeFiles/cmTC_246f2.dir/testCXXCompiler.cxx.o  -o cmTC_246f2   && :
    ninja: build stopped: subcommand failed.

Now it's compiled. :D But (of course there is but :D) when i use ninja instead of make the is a lot of errors and it stops with ninja: build stopped: subcommand failed :(

--- a/packages/opencv/build.sh
+++ b/packages/opencv/build.sh
@@ -2,9 +2,13 @@ TERMUX_PKG_HOMEPAGE=https://opencv.org/
 TERMUX_PKG_DESCRIPTION="Open Source Computer Vision Library"
 TERMUX_PKG_LICENSE="Apache-2.0"
 TERMUX_PKG_MAINTAINER="@termux"
-TERMUX_PKG_VERSION=4.5.5
-TERMUX_PKG_SRCURL=https://github.com/opencv/opencv/archive/${TERMUX_PKG_VERSION}.tar.gz
-TERMUX_PKG_SHA256=a1cfdcf6619387ca9e232687504da996aaa9f7b5689986b8331ec02cb61d28ad
+TERMUX_PKG_VERSION=(4.5.5)
+TERMUX_PKG_VERSION+=(1.22.3) # NumPy version
+TERMUX_PKG_REVISION=1
+TERMUX_PKG_SRCURL=(https://github.com/opencv/opencv/archive/${TERMUX_PKG_VERSION}.tar.gz
+                   https://github.com/numpy/numpy/archive/refs/tags/v${TERMUX_PKG_VERSION[1]}.tar.gz)
+TERMUX_PKG_SHA256=(a1cfdcf6619387ca9e232687504da996aaa9f7b5689986b8331ec02cb61d28ad
+                   c8f3ec591e3f17b939220f2b9eabb4c5e2db330f8af62c0a3aeee8a4d1a6c0db)
 TERMUX_PKG_DEPENDS="libc++, libjpeg-turbo, libpng, libprotobuf, libtiff, libwebp, openjpeg, openjpeg-tools, zlib"
 TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
 -DANDROID_NO_TERMUX=OFF
@@ -12,6 +16,10 @@ TERMUX_PKG_EXTRA_CONFIGURE_ARGS="
 -DBUILD_PROTOBUF=OFF
 -DPROTOBUF_UPDATE_FILES=ON
 "
+
+termux_step_post_get_source() {
+	mv numpy-${TERMUX_PKG_VERSION[1]} numpy
+}
 
 termux_step_pre_configure() {
 	termux_setup_protobuf
@@ -35,6 +43,27 @@ termux_step_pre_configure() {
 		_NEED_DUMMY_LIBRT_A=true
 		echo '!<arch>' > $_LIBRT_A
 	fi
+
+	# For Python bindings
+	_PYTHON_VERSION=$(. $TERMUX_SCRIPTDIR/packages/python/build.sh; echo $_MAJOR_VERSION)
+	termux_setup_python_crossenv
+	pushd $TERMUX_PYTHON_CROSSENV_SRCDIR
+	_CROSSENV_PREFIX=$TERMUX_PKG_BUILDDIR/python-crossenv-prefix
+	python${_PYTHON_VERSION} -m crossenv \
+		$TERMUX_PREFIX/bin/python${_PYTHON_VERSION} \
+		${_CROSSENV_PREFIX}
+	popd
+	. ${_CROSSENV_PREFIX}/bin/activate
+	LDFLAGS+=" -lpython${_PYTHON_VERSION}"
+	export NPY_DISABLE_SVML=1
+	pushd $TERMUX_PKG_SRCDIR/numpy
+	pip install .
+	popd
+	TERMUX_PKG_EXTRA_CONFIGURE_ARGS+="
+		-DPYTHON_DEFAULT_EXECUTABLE=python
+		-DPYTHON3_INCLUDE_PATH=$TERMUX_PREFIX/include/python${_PYTHON_VERSION}
+		-DPYTHON3_NUMPY_INCLUDE_DIRS=${_CROSSENV_PREFIX}/cross/lib/python${_PYTHON_VERSION}/site-packages/numpy/core/include
+		"
 }
 
 termux_step_post_make_install() {

--- a/packages/opencv/opencv-python.subpackage.sh
+++ b/packages/opencv/opencv-python.subpackage.sh
@@ -1,0 +1,8 @@
+TERMUX_SUBPKG_INCLUDE="lib/python*"
+TERMUX_SUBPKG_DESCRIPTION="Python bindings for OpenCV"
+TERMUX_SUBPKG_DEPENDS="python"
+
+termux_step_create_subpkg_debscripts() {
+	echo "#!$TERMUX_PREFIX/bin/sh" > postinst
+	echo "pip3 install numpy" >> postinst
+}

